## Google Capstone Project : How Can a Wellness Technology Company Play It Smart?

## ASK 

1. How non Bellabeat users use smart devices (find trends)?
2. How to apply these trends in Bellabeat products?
3. Recommend strategies to market product?

<strong>Business Task: Using data to influence market strategy</strong>


## PREPARE

1. Where is your data stored?
<p>Ans: Kaggle

2. How is the data organised?
<p>Ans: In both formats: Wide and Long

3. Are there any issues with bias or credibility in this data? (Is data ROCCC)
<p>Ans: Reliable: No, as data has low number of users
     <p>Original: No, as it is from third party sources
     <p>Cited: Yes
     <p>Comprehensible: Some match with Bellabeat features
     <p>Current: No, as it is from 2016 but will not affect grealy with user habits  in 2021

4. Are there any problems with the data?
<p>Ans: Data cannot be use for any business recommendation as it is not ROCCC.



In [39]:
#importing packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import ipywidgets as widgets
import plotly.express as px

In [25]:
#import csv files
daily_activity_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/dailyActivity_merged.csv")
heartrate_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/heartrate_seconds_merged.csv")
sleepday_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/sleepDay_merged.csv")
hourly_calories_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/hourlyCalories_merged.csv")
hourly_steps_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/hourlySteps_merged.csv")
hourly_intensities_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/hourlyIntensities_merged.csv")
minute_calories_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/minuteCaloriesNarrow_merged.csv")
minute_intensities_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/minuteIntensitiesNarrow_merged.csv")
minute_sleep_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/minuteSleep_merged.csv")
minute_steps_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/minuteStepsNarrow_merged.csv")
minute_MET_df = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/minuteMETsNarrow_merged.csv")
weight_log = pd.read_csv("/content/drive/MyDrive/Fitness Tracker Capstone/weightLogInfo_merged.csv")
dataframe_list = [daily_activity_df, heartrate_df, sleepday_df, hourly_calories_df, hourly_steps_df, hourly_intensities_df
                  , minute_calories_df, minute_intensities_df, minute_sleep_df, minute_steps_df, minute_MET_df, weight_log]
dataframe_name_list = ["daily_activity_df", "heartrate_df", "sleepday_df", "hourly_calories_df", "hourly_steps_df", "hourly_intensities_df"
                  , "minute_calories_df", "minute_intensities_df", "minute_sleep_df", "minute_steps_df", "minute_MET_df", "weight_log"]

In [26]:
#Viewing the dataframes
for i, en in enumerate(dataframe_list):
  print("For Dataframe",dataframe_name_list[i])
  display(en.head())
  print()
  print()

For Dataframe daily_activity_df


Id ActivityDate  ...  SedentaryMinutes  Calories
0  1503960366    4/12/2016  ...               728      1985
1  1503960366    4/13/2016  ...               776      1797
2  1503960366    4/14/2016  ...              1218      1776
3  1503960366    4/15/2016  ...               726      1745
4  1503960366    4/16/2016  ...               773      1863

[5 rows x 15 columns]



For Dataframe heartrate_df


Id                  Time  Value
0  2022484408  4/12/2016 7:21:00 AM     97
1  2022484408  4/12/2016 7:21:05 AM    102
2  2022484408  4/12/2016 7:21:10 AM    105
3  2022484408  4/12/2016 7:21:20 AM    103
4  2022484408  4/12/2016 7:21:25 AM    101



For Dataframe sleepday_df


Id               SleepDay  ...  TotalMinutesAsleep  TotalTimeInBed
0  1503960366  4/12/2016 12:00:00 AM  ...                 327             346
1  1503960366  4/13/2016 12:00:00 AM  ...                 384             407
2  1503960366  4/15/2016 12:00:00 AM  ...                 412             442
3  1503960366  4/16/2016 12:00:00 AM  ...                 340             367
4  1503960366  4/17/2016 12:00:00 AM  ...                 700             712

[5 rows x 5 columns]



For Dataframe hourly_calories_df


Id           ActivityHour  Calories
0  1503960366  4/12/2016 12:00:00 AM        81
1  1503960366   4/12/2016 1:00:00 AM        61
2  1503960366   4/12/2016 2:00:00 AM        59
3  1503960366   4/12/2016 3:00:00 AM        47
4  1503960366   4/12/2016 4:00:00 AM        48



For Dataframe hourly_steps_df


Id           ActivityHour  StepTotal
0  1503960366  4/12/2016 12:00:00 AM        373
1  1503960366   4/12/2016 1:00:00 AM        160
2  1503960366   4/12/2016 2:00:00 AM        151
3  1503960366   4/12/2016 3:00:00 AM          0
4  1503960366   4/12/2016 4:00:00 AM          0



For Dataframe hourly_intensities_df


Id           ActivityHour  TotalIntensity  AverageIntensity
0  1503960366  4/12/2016 12:00:00 AM              20          0.333333
1  1503960366   4/12/2016 1:00:00 AM               8          0.133333
2  1503960366   4/12/2016 2:00:00 AM               7          0.116667
3  1503960366   4/12/2016 3:00:00 AM               0          0.000000
4  1503960366   4/12/2016 4:00:00 AM               0          0.000000



For Dataframe minute_calories_df


Id         ActivityMinute  Calories
0  1503960366  4/12/2016 12:00:00 AM    0.7865
1  1503960366  4/12/2016 12:01:00 AM    0.7865
2  1503960366  4/12/2016 12:02:00 AM    0.7865
3  1503960366  4/12/2016 12:03:00 AM    0.7865
4  1503960366  4/12/2016 12:04:00 AM    0.7865



For Dataframe minute_intensities_df


Id         ActivityMinute  Intensity
0  1503960366  4/12/2016 12:00:00 AM          0
1  1503960366  4/12/2016 12:01:00 AM          0
2  1503960366  4/12/2016 12:02:00 AM          0
3  1503960366  4/12/2016 12:03:00 AM          0
4  1503960366  4/12/2016 12:04:00 AM          0



For Dataframe minute_sleep_df


Id                  date  value        logId
0  1503960366  4/12/2016 2:47:30 AM      3  11380564589
1  1503960366  4/12/2016 2:48:30 AM      2  11380564589
2  1503960366  4/12/2016 2:49:30 AM      1  11380564589
3  1503960366  4/12/2016 2:50:30 AM      1  11380564589
4  1503960366  4/12/2016 2:51:30 AM      1  11380564589



For Dataframe minute_steps_df


Id         ActivityMinute  Steps
0  1503960366  4/12/2016 12:00:00 AM      0
1  1503960366  4/12/2016 12:01:00 AM      0
2  1503960366  4/12/2016 12:02:00 AM      0
3  1503960366  4/12/2016 12:03:00 AM      0
4  1503960366  4/12/2016 12:04:00 AM      0



For Dataframe minute_MET_df


Id         ActivityMinute  METs
0  1503960366  4/12/2016 12:00:00 AM    10
1  1503960366  4/12/2016 12:01:00 AM    10
2  1503960366  4/12/2016 12:02:00 AM    10
3  1503960366  4/12/2016 12:03:00 AM    10
4  1503960366  4/12/2016 12:04:00 AM    10



For Dataframe weight_log


Id                   Date  ...  IsManualReport          LogId
0  1503960366   5/2/2016 11:59:59 PM  ...            True  1462233599000
1  1503960366   5/3/2016 11:59:59 PM  ...            True  1462319999000
2  1927972279   4/13/2016 1:08:52 AM  ...           False  1460509732000
3  2873212765  4/21/2016 11:59:59 PM  ...            True  1461283199000
4  2873212765  5/12/2016 11:59:59 PM  ...            True  1463097599000

[5 rows x 8 columns]

In [27]:
# Checking the types of each column in the dataframe
for i, en in enumerate(dataframe_list):
  print("For Dataframe: ",dataframe_name_list[i])
  print()
  print(en.dtypes)
  print()
  print()

For Dataframe:  daily_activity_df

Id                            int64
ActivityDate                 object
TotalSteps                    int64
TotalDistance               float64
TrackerDistance             float64
LoggedActivitiesDistance    float64
VeryActiveDistance          float64
ModeratelyActiveDistance    float64
LightActiveDistance         float64
SedentaryActiveDistance     float64
VeryActiveMinutes             int64
FairlyActiveMinutes           int64
LightlyActiveMinutes          int64
SedentaryMinutes              int64
Calories                      int64
dtype: object


For Dataframe:  heartrate_df

Id        int64
Time     object
Value     int64
dtype: object


For Dataframe:  sleepday_df

Id                     int64
SleepDay              object
TotalSleepRecords      int64
TotalMinutesAsleep     int64
TotalTimeInBed         int64
dtype: object


For Dataframe:  hourly_calories_df

Id               int64
ActivityHour    object
Calories         int64
dtype: object


For 

## Process

###  Cleaning the dataset
Documenting the cleaning process using the ipython notebook

In [28]:
#Checking for Nan Values
for i, en in enumerate(dataframe_list):
  if en.isna().sum().sum() > 0:
    print("Nan Values for Dataframe",dataframe_name_list[i],"is",en.isna().sum().sum())

Nan Values for Dataframe weight_log is 65


In [29]:
#Finding duplicates in dataframe
for i, en in enumerate(dataframe_list):
  if en.drop_duplicates().shape[0] != en.shape[0]:
    print("Number of Duplicates for Dataframe",dataframe_name_list[i],"is",en.shape[0] - en.drop_duplicates().shape[0])
sleepday_df = sleepday_df.drop_duplicates()
minute_sleep_df = minute_sleep_df.drop_duplicates()

Number of Duplicates for Dataframe sleepday_df is 3
Number of Duplicates for Dataframe minute_sleep_df is 543


In [30]:
# As the number of participants mentioned were 33 in the metadata of the data, we are verifying the same
for i, en in enumerate(dataframe_list):
  print("For Dataframe",dataframe_name_list[i],":",en["Id"].unique().shape[0])

For Dataframe daily_activity_df : 33
For Dataframe heartrate_df : 14
For Dataframe sleepday_df : 24
For Dataframe hourly_calories_df : 33
For Dataframe hourly_steps_df : 33
For Dataframe hourly_intensities_df : 33
For Dataframe minute_calories_df : 33
For Dataframe minute_intensities_df : 33
For Dataframe minute_sleep_df : 24
For Dataframe minute_steps_df : 33
For Dataframe minute_MET_df : 33
For Dataframe weight_log : 8


We are not doing any further data analysis on weight_log and heartrate as the sample size is small as compared to number of users registered for survey while others are matching the number of participants except of sleepday_df and minute_sleep_df. But, we are not excluding the sleepday_df and minute_sleep_df as the sample size is 24 which is okay(in this case).

## Analyze and Share

In [31]:
daily_activity_df.columns

Index(['Id', 'ActivityDate', 'TotalSteps', 'TotalDistance', 'TrackerDistance',
       'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'Calories'],
      dtype='object')

In [32]:
daily_activity_df.iloc[:,1:].describe()

TotalSteps  TotalDistance  ...  SedentaryMinutes     Calories
count    940.000000     940.000000  ...        940.000000   940.000000
mean    7637.910638       5.489702  ...        991.210638  2303.609574
std     5087.150742       3.924606  ...        301.267437   718.166862
min        0.000000       0.000000  ...          0.000000     0.000000
25%     3789.750000       2.620000  ...        729.750000  1828.500000
50%     7405.500000       5.245000  ...       1057.500000  2134.000000
75%    10727.000000       7.712500  ...       1229.500000  2793.250000
max    36019.000000      28.030001  ...       1440.000000  4900.000000

[8 rows x 13 columns]

We infer the following points from the above data and follwing tracks leading to further analysing and sharing of data:-
1. The mean total steps taken per day is 7638 and 50% of data is less than 8000 as per a study, 8000 steps are to be taken a day to reduce health risks related to death considerably.<a href="https://www.nih.gov/news-events/nih-research-matters/number-steps-day-more-important-step-intensity">[Source]</a>. The minimum steps is 0. We can see that activity minutes are very low. We can sum them up together and check if they come up for 20 minutes a day (excluding sedentary activity)<a href="https://theconversation.com/how-much-do-sedentary-people-really-need-to-move-its-less-than-you-think-114824">[Source]</a>.
2. Logged activity distance are very low.
3. Sedentary minutes are very high and thus, can be an influencing parameter in calories burning. 
4. We can also track for week day and weekends calories data.

In [33]:
sleepday_df.iloc[:,1:].describe()

TotalSleepRecords  TotalMinutesAsleep  TotalTimeInBed
count         410.000000          410.000000      410.000000
mean            1.119512          419.173171      458.482927
std             0.346636          118.635918      127.455140
min             1.000000           58.000000       61.000000
25%             1.000000          361.000000      403.750000
50%             1.000000          432.500000      463.000000
75%             1.000000          490.000000      526.000000
max             3.000000          796.000000      961.000000

We infer the follwoing points and tracks from the above table:-
1. The maximum sleep hours in the data is 16 hours. Either there is fault in sensor or user is really sleeping for long durations. Recommended sleep hours is 7-9 hours for adults.
2. Almost, 75% of the data is within the recommended limit.

#### Working on 1st Point



In [80]:
#First, checking for days where active minutes were less than 20 mins
daily_activity_df["ActiveMinutes"] = daily_activity_df["LightlyActiveMinutes"] + daily_activity_df["FairlyActiveMinutes"] + daily_activity_df["VeryActiveMinutes"]
px.scatter(daily_activity_df[daily_activity_df["ActiveMinutes"] < 20],"ActiveMinutes","Calories", title="Calories burned Vs Activity Minutes")

In [77]:
#Checking further for anu inflence of sedentary activity
px.scatter(daily_activity_df[daily_activity_df["ActiveMinutes"] < 20],"SedentaryMinutes","Calories", title="Calories burned by activity less than 20 mins")

Above is the data for the calories burned for an active activity of less than 20 mins. We can see that it is highly influenced by sedentary activity.

#### Working on 2nd Point

In [143]:
#We will now check for those entries where logged activity distance is greater than zero
daily_activity_df[daily_activity_df["LoggedActivitiesDistance"] > 0].shape[0]

32

From here, we can say that either users do not tick after a challenge set by the admin is complete or the engagement is quite low.

#### Working on 3rd Point

Now, we will check for calories Vs Steps to see any trends from the data

In [149]:
px.scatter(daily_activity_df,"SedentaryMinutes","TotalSteps")

As the above plot is showing no trends and from this, we can see that tracking steps for the entire date at once can be not prove useful as for the same steps, the minutes are high.

In [161]:
px.scatter(daily_activity_df,"TotalSteps","Calories")

### Act

From the analysis, we recommend the following changes :-


1. We can encourage users to minimum walk 8000 steps a day, 20 mins exercise a day or a combination of both. For this, we can make a setting of setting daily target adjuestable by the user and gamify it by including a game of growing a pet or similar game where steps would be acting as coins which would be used for in-purchases in game to encourage more walking and exercise.
2. To make it more interesting, they can share their scores with ther users and region leaderboards can be set up which they can view in their apps for an healthy competition. 
3. For the display of result of steps taken in a day, it can be shown in the mail or personal dashboards in an amusing way like "Congratulations! You have taken ### steps today. You are taking a step towards saving the environment by saving ### amount of carbon/fuel emmission if you were using a vehicle. A small steps taken by you is a giant leap for saving the environment!"<a href="https://www.lowa.co.uk/lovelowa/74-six-environmental-benefits-of-walking#:~:text=Research%20suggests%20choosing%20to%20walk,emissions%20by%2086kg%20a%20year.">[Source]</a>. <br>Same can be done for exercise activities.
4. We can encourage users by giving reminders to tick off any logged in challenges they have taken and also, keep an monthly challenge for all users on reigonal basis.
5. As sedentary activities influence highly on calorific value, we can give alerts to users when they reach off the sedentary limit.